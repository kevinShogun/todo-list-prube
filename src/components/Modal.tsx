
interface ModalProps {
    children: React.ReactNode;
    id?: string;
}

export const Modal = (
    { children, id = 'my_modal_2' }: ModalProps,
) => {

    return (
        <dialog id={id} className="modal">
            <div className="modal-box">
                {children}
            </div>
            <form method="dialog" className="modal-backdrop">
                <button>close</button>
            </form>
        </dialog>
    )
}
