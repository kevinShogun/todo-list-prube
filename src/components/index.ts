export { TodoScreen } from "./TodoScreen";
export { SideBar } from "./SideBar";

