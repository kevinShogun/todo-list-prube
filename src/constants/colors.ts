export const COLORS = [ "#d2ceff", "#ffdbdb", "#d1e5f7", "#ffcece", "#daf2d6"]