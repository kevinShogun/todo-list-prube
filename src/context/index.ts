export { TodoProvider } from "./TodoProvider";
export { TodoReducer } from "./TodoReducer";
export { TodoContext } from "./TodoContext";


