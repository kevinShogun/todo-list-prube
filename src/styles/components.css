.button-text{
    @apply font-semibold text-lg md:text-xl lg:text-2xl;
}

.color-circle{
    @apply w-5 h-5 shadow-md rounded-full cursor-pointer mx-1 md:w-7 md:h-7 tooltip;
}

.input-text{
    @apply input input-bordered w-full mt-2;
}

form{
    @apply flex flex-col gap-3 select-none;
}
.textarea-form{
    @apply textarea textarea-bordered w-full  mt-2 resize-none min-h-32;    
}
