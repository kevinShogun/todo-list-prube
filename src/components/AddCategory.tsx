import { useState, useContext, useEffect } from 'react'
import { v4 as uuid } from 'uuid';
import { TodoContext } from '../context';
import { toggleModal } from '../helper/modalAction';
import { Category } from '../interfaces'
import { COLORS } from '../constants/colors';
import { Modal } from './Modal'
import '../styles/components.css'


/**
 * @description This code defines a React functional component called AddCategory
 * that is responsible for adding and updating categories. 
 * It uses the useState and useContext hooks from React to manage the state and access the context. 
 * The component renders a button that opens a modal dialog when clicked. 
 * Inside the modal, the user can enter a category name and select a color from a predefined list. 
 * The component also listens for changes in the currentCategory context variable and 
 * updates the input fields accordingly.
 * 
 * @returns  {JSX.Element} The rendered todo item.
 */
export const AddCategory = () => {

    const [categoryName, setCategoryName] = useState('');
    const [color, setColor] = useState('');

    const { addCategory, currentCategory, replaceCategory } = useContext(TodoContext);

    useEffect(() => {
        if (currentCategory) {
            setCategoryName(currentCategory.name);
            setColor(currentCategory.color);
            toggleModal({ value: true });
        }
    }, [currentCategory])


    const handleCategory = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault()

        if (categoryName.trim() === '' || color.trim() === '') {
            return
        }

        const newCategory: Category = {
            id: uuid(),
            name: categoryName,
            color
        }

        if (currentCategory) {
            const updatedCategory = { ...currentCategory, name: categoryName, color }
            replaceCategory(updatedCategory)
        }else{
            addCategory(newCategory);
        }
        
        setCategoryName('');
        setColor('');

        toggleModal({ value: false });
    }

    return (
        <div>
            <button className="btn btn-ghost"
                onClick={() => {
                    toggleModal({ value: true });
                }}
            >
                Agregar una categoria
                <i className='bx bx-plus'></i>
            </button>
            <Modal>
                <h2 className="button-text">
                    Agregar una categoria
                </h2>
                <div className="divider" />
                <form method="dialog" className="modal-form" onSubmit={handleCategory}>
                    <div className="form-control">
                        <label htmlFor="category">Categoria: </label>
                        <input type="text" id="category" className='input-text'
                            value={categoryName}
                            onChange={(e) => setCategoryName(e.target.value)}
                        />
                    </div>

                    <div className="form-control">
                        <label htmlFor="color">Color: </label>
                        <div className='flex gap-3'>
                            {COLORS.map((c, index) => (
                                <div
                                    key={index}
                                    className={`color-circle ${color === c && 'border-2 border-black'}`}
                                    style={{ backgroundColor: c }}
                                    onClick={() => setColor(c)}
                                />

                            ))}

                        </div>

                    </div>
                    <button type="submit" className="btn btn-primary">
                        {currentCategory ? 'Actualizar' : 'Agregar'}
                    </button>
                </form>
            </Modal>
        </div>
    )
}
