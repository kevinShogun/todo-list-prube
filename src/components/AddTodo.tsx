import { useState, useContext, useEffect } from 'react';
import { v4 as uuid } from 'uuid';
import { TodoContext } from '../context';
import { Category, Todo } from '../interfaces';
import { toggleModal } from "../helper/modalAction"
import { Modal } from "./Modal"
import '../styles/components.css'

type NewTodo = {
    title: string;
    body: string;
    categories: Category[];
}

/**
 * @description This code defines a React component called AddTodo that is responsible for adding a new todo item. 
 * It uses the useState and useContext hooks from React to manage the state and access the context. 
 * The component renders a button that opens a modal dialog when clicked. 
 * Inside the modal, the user can enter a title, description, and select categories for the new todo item. 
 * The component also handles updating an existing todo item if one is provided through the context.
 * @returns {JSX.Element}
 */

export const AddTodo = () => {

    const [todo, settodo] = useState<NewTodo>({
        title: '',
        body: '',
        categories: []
    });
    const { categories, addTodo, replaceTodo, currentTodo} = useContext(TodoContext);


    useEffect(() => {
        if (currentTodo) {
            settodo({
                title: currentTodo.title,
                body: currentTodo.body,
                categories: currentTodo.categories
            });
            toggleModal({ value: true, id: 'add_todo'});
        }
    }, [currentTodo])


    const setCategories = (c: Category) => {

        if(todo.categories.includes(c)){
            const newCategories = todo.categories.filter((cat) => cat.id !== c.id);
            return settodo({ ...todo, categories: newCategories })
        }else{
            return settodo({ 
                ...todo, 
                categories: [...todo.categories, c]
            })
        }
    }

    const handleAddTodo = () => {
        const newTodo:Todo = {
            ...todo,
            id: uuid(),
        }

        if(currentTodo){
            const updatedTodo = {
                ...currentTodo,
                title: todo.title,
                body: todo.body,
                categories: todo.categories
            }
            replaceTodo(updatedTodo);
        }else{
            addTodo(newTodo);
        }

        settodo({
            title: '',
            body: '',
            categories: []
        });
        toggleModal({ value: false, id: 'add_todo'});

    }

    return (
        <div>
            <button className="btn btn-ghost"
                onClick={() => {
                    toggleModal({ value: true, id: 'add_todo' });
                }}
            >
                Agregar una tarea
                <i className='bx bx-plus'></i>
            </button>
            <br /> <br />
            <Modal
                id="add_todo"
            >
                <h2 className="button-text">
                    Nueva tarea
                </h2>
                <div className="divider" />
                <form method="dialog" className="modal-form" onSubmit={handleAddTodo}>
                    <div className="form-control">
                        <label htmlFor="title">Título: </label>
                        <input type="text" id="title" className='input-text'
                            value={todo.title}
                            onChange={(e) => settodo({ ...todo, title: e.target.value })}
                        />
                    </div>

                    <div className="form-control">
                        <label htmlFor="description">Descripción: </label>
                        <textarea
                            id="description"
                            value={todo.body}
                            className='textarea-form'
                            onChange={(e) => settodo({ ...todo, body: e.target.value })}
                        ></textarea>
                    </div>
                    <div className="form-control">
                        <label htmlFor="color">Categorías: </label>
                        <div className='flex gap-3 mt-2'>
                            {categories?.map((c, index) => (
                                <div
                                    key={index}
                                    className={`color-circle ${todo.categories.includes(c) ? 'border-2 border-black' : ''}`}
                                    data-tip={c.name}
                                    style={{ backgroundColor: c.color }}
                                    onClick={() =>  setCategories(c) }
                                />

                            ))}

                        </div>

                    </div>

                    <button type="submit" className="btn btn-primary">
                        {currentTodo ? 'Actualizar' : 'Agregar'}
                    </button>
                </form>
            </Modal>
        </div>
    )
}
