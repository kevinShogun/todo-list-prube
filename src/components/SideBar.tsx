import { useContext } from 'react'
import { MagicMotion } from "react-magic-motion";
import { AddCategory } from "./AddCategory"
import { TodoContext } from "../context";

export const SideBar = () => {

  const { categories, setCurrentCategory, removeCategory } = useContext(TodoContext);

  return (
    <aside className="sidebar">
      <h2 className="font-semibold text-2xl md:text-3xl lg:text-5xl">
        todo
      </h2>

      <MagicMotion>
        <div>
          <ul className="mt-5">
            {categories.map((category) => (
              <li key={category.id} className="list-item">
                <div className='flex w-3/4 items-center gap-2'>
                  <span className="category-color" style={{ backgroundColor: category.color }} />
                  <span className='list-item'>{category.name}</span>
                </div>
                <div className="dropdown dropdown-end">
                  <div tabIndex={0} role="button">
                    <i className='bx bx-dots-horizontal-rounded'></i>
                  </div>
                  <ul tabIndex={0} className="dropdown-item">
                    <li>
                      <a onClick={() => setCurrentCategory(category)}> Editar </a>
                    </li>
                    <li>
                      <a onClick={() => removeCategory(category.id)}> Eliminar </a>
                    </li>
                  </ul>
                </div>
              </li>
            ))}
          </ul>
        </div>
      </MagicMotion>

      <div className="divider" />
      <AddCategory />
      <img
        src='./undraw_to_do_re_jaef.svg'
        alt="note list"
        className="w-3/4 mx-auto mt-5 hidden lg:block"
      />
    </aside>
  )
}
/***
 * Edad de 25 a 35 a√±os
talentohumano.ni@grupolgb.com
whatsapp 76221311
 */
