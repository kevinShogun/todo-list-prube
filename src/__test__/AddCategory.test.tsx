import { fireEvent, render, screen } from "@testing-library/react";
import '@testing-library/jest-dom'
import { COLORS } from "../constants/colors";
import { AddCategory } from "../components/AddCategory";
import { TodoContext } from "../context";

describe("AddCategory", () => {

	//A snapshot test
	test("should match snapshot", () => {
		const component = render(<AddCategory />);
		expect(component).toMatchSnapshot();
	});

	test("should render a button that opens a modal dialog when clicked", async () => {
		// find the button by class name
		render(<AddCategory />);
		const button = screen.getByText(/Agregar una categoria/i);
		// click the button
		button.click();
		// find the modal by class name
		const modal = screen.getByText(/Nueva categoria/i);
		// assert the modal is visible
		expect(modal).toBeVisible();
	});

	it('should render a modal dialog with input fields for category name and color selection', () => {

		render(<AddCategory />);
		const button = screen.getByText(/Agregar una categoria/i);
		fireEvent.click(button);

		const modal = screen.getByRole('dialog');
		expect(modal).toBeVisible();

		const categoryInput = screen.getByLabelText('Categoria:');
		expect(categoryInput).toBeInTheDocument();

		const colorCircles = modal.querySelectorAll('.color-circle');
		// Assert that the color selection circles are rendered
		expect(colorCircles.length).toBe(COLORS.length);

	});

	test('should not submit the form if the category name or color is empty', () => {
		render(<AddCategory />);
		// Find the button element and click it to open the modal dialog
		const button = screen.getByText(/Agregar una categoria/i);
		fireEvent.click(button);

		// Find the form element
		const form = document.querySelector('form');
		if (!form) return;

		// Find the submit button element
		const submitButton = screen.getByText('Agregar');
		fireEvent.click(submitButton);

		// Submit the form without entering any values
		fireEvent.submit(form);

		// Assert that the form was not submitted
		expect(submitButton).toBeVisible();
	});

	test('should replace "Agregar" button with "Actualizar" button in the modal when current category is not null', () => {

		// delete the las instance of AddCategory
		
		
		const FC = render(
			<TodoContext.Provider
				value={{
					todos: [],
					categories: [],
					currentTodo: null,
					setCurrentTodo: () => {}, // Add the missing properties here
					currentCategory: { id: '1', name: 'Work', color: 'red' },
					addCategory: () => {},
					replaceCategory: () => {},
					addTodo: () => {},
					replaceTodo: () => {},
					removeCategory(id) {},
					removeTodo: () => {},
					setCurrentCategory(category) {},
					toggleTodo: () => {},
					// Add the remaining properties here
				}}				
			>
				<AddCategory/>
			</TodoContext.Provider>
		);

		// Find the button element and click it to open the modal dialog
		const button = FC.getAllByText(/Agregar una categoria/i);
		fireEvent.click(button[0]);

		const modal = FC.getAllByRole('dialog');
		expect(modal[0]).toBeVisible();

		const submitButton = FC.getByText('Actualizar');
		expect(submitButton).toBeVisible();
	});


});
