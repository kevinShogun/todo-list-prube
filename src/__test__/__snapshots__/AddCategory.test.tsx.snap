// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddCategory > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <button
          class="btn btn-ghost"
        >
          Agregar una categoria
          <i
            class="bx bx-plus"
          />
        </button>
        <dialog
          class="modal"
          id="my_modal_2"
        >
          <div
            class="modal-box"
          >
            <h2
              class="button-text"
            >
              Nueva categoria
            </h2>
            <div
              class="divider"
            />
            <form
              class="modal-form"
              method="dialog"
            >
              <div
                class="form-control"
              >
                <label
                  for="category"
                >
                  Categoria: 
                </label>
                <input
                  class="input-text"
                  id="category"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-control"
              >
                <label
                  for="color"
                >
                  Color: 
                </label>
                <div
                  class="flex gap-3"
                >
                  <div
                    class="color-circle false"
                    style="background-color: #d2ceff;"
                  />
                  <div
                    class="color-circle false"
                    style="background-color: #ffdbdb;"
                  />
                  <div
                    class="color-circle false"
                    style="background-color: #d1e5f7;"
                  />
                  <div
                    class="color-circle false"
                    style="background-color: #ffcece;"
                  />
                  <div
                    class="color-circle false"
                    style="background-color: #daf2d6;"
                  />
                </div>
              </div>
              <button
                class="btn btn-primary"
                type="submit"
              >
                Agregar
              </button>
            </form>
          </div>
          <form
            class="modal-backdrop"
            method="dialog"
          >
            <button>
              close
            </button>
          </form>
        </dialog>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <button
        class="btn btn-ghost"
      >
        Agregar una categoria
        <i
          class="bx bx-plus"
        />
      </button>
      <dialog
        class="modal"
        id="my_modal_2"
      >
        <div
          class="modal-box"
        >
          <h2
            class="button-text"
          >
            Nueva categoria
          </h2>
          <div
            class="divider"
          />
          <form
            class="modal-form"
            method="dialog"
          >
            <div
              class="form-control"
            >
              <label
                for="category"
              >
                Categoria: 
              </label>
              <input
                class="input-text"
                id="category"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-control"
            >
              <label
                for="color"
              >
                Color: 
              </label>
              <div
                class="flex gap-3"
              >
                <div
                  class="color-circle false"
                  style="background-color: #d2ceff;"
                />
                <div
                  class="color-circle false"
                  style="background-color: #ffdbdb;"
                />
                <div
                  class="color-circle false"
                  style="background-color: #d1e5f7;"
                />
                <div
                  class="color-circle false"
                  style="background-color: #ffcece;"
                />
                <div
                  class="color-circle false"
                  style="background-color: #daf2d6;"
                />
              </div>
            </div>
            <button
              class="btn btn-primary"
              type="submit"
            >
              Agregar
            </button>
          </form>
        </div>
        <form
          class="modal-backdrop"
          method="dialog"
        >
          <button>
            close
          </button>
        </form>
      </dialog>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
